eZ/Publish/SPI/Limitation/Target/Version.php

```diff
@@ -0,0 +1,185 @@
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\SPI\Limitation\Target;
+
+use eZ\Publish\API\Repository\Values\Content\Content;
+use eZ\Publish\API\Repository\Values\Content\Field;
+use eZ\Publish\API\Repository\Values\Content\VersionInfo;
+use eZ\Publish\Core\Base\Exceptions\BadStateException;
+use eZ\Publish\Core\Base\Exceptions\InvalidArgumentException;
+use eZ\Publish\SPI\Limitation\Target;
+use eZ\Publish\SPI\Persistence\ValueObject;
+
+/**
+ * Version Limitation target. Indicates an intent to create new Version.
+ *
+ * @property-read string[] $translateToLanguageCodes
+ * @property-read int $newStatus
+ * @property-read string $newInitialLanguageCode
+ * @property-read string[] $updateTranslationsLanguageCodes
+ * @property-read \eZ\Publish\API\Repository\Values\Content\Content|null $lastPublishedContent
+ * @property-read \eZ\Publish\API\Repository\Values\Content\Content|null $targetContent
+ */
+final class Version extends ValueObject implements Target
+{
+    /**
+     * List of language codes of translations.
+
+     * @var string[]
+     */
+    protected $translateToLanguageCodes = [];
+
+    /**
+     * One of the following: STATUS_DRAFT, STATUS_PUBLISHED, STATUS_ARCHIVED.
+     *
+     * @see \eZ\Publish\API\Repository\Values\Content\VersionInfo::STATUS_DRAFT
+     * @see \eZ\Publish\API\Repository\Values\Content\VersionInfo::STATUS_PUBLISHED
+     * @see \eZ\Publish\API\Repository\Values\Content\VersionInfo::STATUS_ARCHIVED
+     *
+     * @var int|null
+     */
+    protected $newStatus;
+
+    /**
+     * Language code of a translation used when updated, can be null for e.g. multiple translations changed.
+     *
+     * @var string|null
+     */
+    protected $newInitialLanguageCode;
+
+    /**
+     * List of language codes of translations to update.
+     *
+     * @var string[]
+     */
+    protected $updateTranslationsLanguageCodes;
+
+    /**
+     * Most recently published content to be used by Limitation for comparison.
+     *
+     * @var \eZ\Publish\API\Repository\Values\Content\Content
+     */
+    protected $lastPublishedContent;
+
+    /**
+     * Target Content to be used by Limitation for comparision.
+     *
+     * @var \eZ\Publish\API\Repository\Values\Content\Content
+     */
+    protected $targetContent;
+
+    /**
+     * Set intent to translate, to an unspecified (yet) language, any from the given list.
+     *
+     * @param array $languageCodes
+     *
+     * @return self
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    public function translateToAnyLanguageOf(array $languageCodes): self
+    {
+        foreach ($languageCodes as $languageCode) {
+            if (!is_string($languageCode) || empty($languageCode)) {
+                throw new InvalidArgumentException('$languageCodes', 'All language codes should be non-empty strings');
+            }
+        }
+
+        $this->translateToLanguageCodes = $languageCodes;
```

Comment:
IMHO objects of this class should be immutable, so. 

A) Please move methods `translateToAnyLanguageOf `, `changeStatusTo ` etc. to separate dedicated builder class e.g. `VersionTargetBuilder`. 
B) Return new instance of `Version` instead of `$this`

Reply:
> IMHO objects of this class should be immutable, so.
> 
> A) Please move methods `translateToAnyLanguageOf `, `changeStatusTo ` etc. to separate dedicated builder class e.g. `VersionTargetBuilder`.
> B) Return new instance of `Version` instead of `$this`

Fixed in 43a90d6c3 with a little bit of refactoring of names.